[Flaky test] TestTemporalityStartTimes flaky on Windows

### Component(s)

cmd/telemetrygen

### Describe the issue you're reporting

e.g. http://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/14301595078/job/40076840440

```
✖  pkg/metrics (455ms)

=== Failed
=== FAIL: pkg/metrics TestTemporalityStartTimes/Delta_temporality_has_different_start_timestamps (0.00s)
    worker_test.go:497: 
make[2]: *** [../../Makefile.Common:103: test] Error 1
        	Error Trace:	D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/cmd/telemetrygen/pkg/metrics/worker_test.go:497
        	            				D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/cmd/telemetrygen/pkg/metrics/worker_test.go:535
        	Error:      	Should be true
        	Test:       	TestTemporalityStartTimes/Delta_temporality_has_different_start_timestamps
make[1]: *** [Makefile:261: cmd/telemetrygen] Error 2
        	Messages:   	delta metrics should have increasing start times
make[1]: *** Waiting for unfinished jobs....

=== FAIL: pkg/metrics TestTemporalityStartTimes (0.00s)

=== FAIL: pkg/metrics TestTemporalityStartTimes/Delta_temporality_has_different_start_timestamps (re-run 1) (0.00s)
    worker_test.go:497: 
        	Error Trace:	D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/cmd/telemetrygen/pkg/metrics/worker_test.go:497
        	            				D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/cmd/telemetrygen/pkg/metrics/worker_test.go:535
        	Error:      	Should be true
        	Test:       	TestTemporalityStartTimes/Delta_temporality_has_different_start_timestamps
        	Messages:   	delta metrics should have increasing start times

=== FAIL: pkg/metrics TestTemporalityStartTimes (re-run 1) (0.00s)

```

--- a/.chloggen/nickange_telemetrygen_metrics_fix-flaky-test.yaml
+++ b/.chloggen/nickange_telemetrygen_metrics_fix-flaky-test.yaml
@@ -1,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: telemetrygen
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Fix flaky test TestTemporalityStartTimes
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [39219]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
+
+# If your change doesn't affect end users or the exported elements of any package,
+# you should instead start your pull request title with [chore] or use the "Skip Changelog" label.
+# Optional: The change log or logs in which this entry should be included.
+# e.g. '[user]' or '[user, api]'
+# Include 'user' if the change is relevant to end users.
+# Include 'api' if there is a change to a library API.
+# Default: '[user]'
+change_logs: []

--- a/cmd/telemetrygen/pkg/metrics/clock.go
+++ b/cmd/telemetrygen/pkg/metrics/clock.go
@@ -1,0 +1,22 @@
+package metrics
+
+import "time"
+
+type Clock interface {
+	Now() time.Time
+}
+
+type realClock struct{}
+
+func (c *realClock) Now() time.Time {
+	return time.Now()
+}
+
+type mockClock struct {
+	now time.Time
+}
+
+func (c *mockClock) Now() time.Time {
+	c.now = c.now.Add(time.Millisecond) // Add 1ms to the mock clock to avoid timestamp collisions
+	return c.now
+}

--- a/cmd/telemetrygen/pkg/metrics/clock.go
+++ b/cmd/telemetrygen/pkg/metrics/clock.go
@@ -1,6 +1,11 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
 package metrics
 
-import "time"
+import (
+	"time"
+)
 
 type Clock interface {
 	Now() time.Time
@@ -17,6 +22,6 @@ type mockClock struct {
 }
 
 func (c *mockClock) Now() time.Time {
-	c.now = c.now.Add(time.Millisecond) // Add 1ms to the mock clock to avoid timestamp collisions
+	c.now = c.now.Add(100 * time.Millisecond) // Add 100ms to the mock clock to avoid timestamp collisions
 	return c.now
 }

--- a/cmd/telemetrygen/pkg/metrics/metrics.go
+++ b/cmd/telemetrygen/pkg/metrics/metrics.go
@@ -77,6 +77,7 @@ func run(c *Config, expF exporterFunc, logger *zap.Logger) error {
 			wg:                     &wg,
 			logger:                 logger.With(zap.Int("worker", i)),
 			index:                  i,
+			clock:                  &realClock{},
 		}
 		exp, err := expF()
 		if err != nil {

--- a/cmd/telemetrygen/pkg/metrics/worker.go
+++ b/cmd/telemetrygen/pkg/metrics/worker.go
@@ -89,8 +89,9 @@ func (w worker) simulateMetrics(res *resource.Resource, exporter sdkmetric.Expor
 	var i int64
 	for w.running.Load() {
 		var metrics []metricdata.Metrics
+		now := time.Now()
 		if w.aggregationTemporality.AsTemporality() == metricdata.DeltaTemporality {
-			startTime = time.Now().Add(-1 * time.Second)
+			startTime = now.Add(-1 * time.Second)
 		}
 		switch w.metricType {
 		case MetricTypeGauge:
@@ -99,7 +100,7 @@ func (w worker) simulateMetrics(res *resource.Resource, exporter sdkmetric.Expor
 				Data: metricdata.Gauge[int64]{
 					DataPoints: []metricdata.DataPoint[int64]{
 						{
-							Time:       time.Now(),
+							Time:       now,
 							Value:      i,
 							Attributes: attribute.NewSet(signalAttrs...),
 							Exemplars:  w.exemplars,
@@ -116,7 +117,7 @@ func (w worker) simulateMetrics(res *resource.Resource, exporter sdkmetric.Expor
 					DataPoints: []metricdata.DataPoint[int64]{
 						{
 							StartTime:  startTime,
-							Time:       time.Now(),
+							Time:       now,
 							Value:      i,
 							Attributes: attribute.NewSet(signalAttrs...),
 							Exemplars:  w.exemplars,
@@ -139,7 +140,7 @@ func (w worker) simulateMetrics(res *resource.Resource, exporter sdkmetric.Expor
 					DataPoints: []metricdata.HistogramDataPoint[int64]{
 						{
 							StartTime:  startTime,
-							Time:       time.Now(),
+							Time:       now,
 							Attributes: attribute.NewSet(signalAttrs...),
 							Exemplars:  w.exemplars,
 							Count:      totalCount,

--- a/cmd/telemetrygen/pkg/metrics/worker.go
+++ b/cmd/telemetrygen/pkg/metrics/worker.go
@@ -29,7 +29,7 @@ type worker struct {
 	wg                     *sync.WaitGroup              // notify when done
 	logger                 *zap.Logger                  // logger
 	index                  int                          // worker index
-	clock                  Clock                        // clock for testing
+	clock                  Clock                        // clock
 }
 
 // We use a 15-element bounds slice for histograms below, so there must be 16 buckets here.

--- a/cmd/telemetrygen/pkg/metrics/worker.go
+++ b/cmd/telemetrygen/pkg/metrics/worker.go
@@ -85,12 +85,12 @@ var histogramBucketSamples = []struct {
 func (w worker) simulateMetrics(res *resource.Resource, exporter sdkmetric.Exporter, signalAttrs []attribute.KeyValue) {
 	limiter := rate.NewLimiter(w.limitPerSecond, 1)
 
-	startTime := time.Now()
+	startTime := w.clock.Now()
 
 	var i int64
 	for w.running.Load() {
 		var metrics []metricdata.Metrics
-		now := time.Now()
+		now := w.clock.Now()
 		if w.aggregationTemporality.AsTemporality() == metricdata.DeltaTemporality {
 			startTime = now.Add(-1 * time.Second)
 		}

--- a/cmd/telemetrygen/pkg/metrics/worker.go
+++ b/cmd/telemetrygen/pkg/metrics/worker.go
@@ -29,6 +29,7 @@ type worker struct {
 	wg                     *sync.WaitGroup              // notify when done
 	logger                 *zap.Logger                  // logger
 	index                  int                          // worker index
+	clock                  Clock                        // clock for testing
 }
 
 // We use a 15-element bounds slice for histograms below, so there must be 16 buckets here.

--- a/cmd/telemetrygen/pkg/metrics/worker_test.go
+++ b/cmd/telemetrygen/pkg/metrics/worker_test.go
@@ -505,6 +505,10 @@ func TestTemporalityStartTimes(t *testing.T) {
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 			m := &mockExporter{}
+			clock := &mockClock{
+				now: time.Now(),
+			}
+
 			running := &atomic.Bool{}
 			running.Store(true)
 
@@ -520,6 +524,7 @@ func TestTemporalityStartTimes(t *testing.T) {
 				limitPerSecond:         rate.Inf,
 				logger:                 zap.NewNop(),
 				wg:                     wg,
+				clock:                  clock,
 			}
 
 			w.simulateMetrics(resource.Default(), m, nil)

--- a/cmd/telemetrygen/pkg/metrics/worker_test.go
+++ b/cmd/telemetrygen/pkg/metrics/worker_test.go
@@ -494,8 +494,9 @@ func TestTemporalityStartTimes(t *testing.T) {
 			name:        "Delta temporality has different start timestamps",
 			temporality: AggregationTemporality(metricdata.DeltaTemporality),
 			checkTimes: func(t *testing.T, firstTime, secondTime time.Time) {
-				assert.True(t, secondTime.After(firstTime),
-					"delta metrics should have increasing start times")
+				diff := secondTime.Sub(firstTime)
+				assert.Greater(t, diff, time.Duration(0),
+					"delta metrics should have increasing start times, got difference: %v", diff)
 			},
 		},
 	}

--- a/cmd/telemetrygen/pkg/metrics/worker_test.go
+++ b/cmd/telemetrygen/pkg/metrics/worker_test.go
@@ -495,8 +495,9 @@ func TestTemporalityStartTimes(t *testing.T) {
 			temporality: AggregationTemporality(metricdata.DeltaTemporality),
 			checkTimes: func(t *testing.T, firstTime, secondTime time.Time) {
 				diff := secondTime.Sub(firstTime)
-				assert.Greater(t, diff, time.Duration(0),
-					"delta metrics should have increasing start times, got difference: %v", diff)
+				t.Logf("Time difference: %v", diff)
+				assert.True(t, secondTime.After(firstTime),
+					"delta metrics should have increasing start times")
 			},
 		},
 	}
@@ -535,14 +536,12 @@ func TestTemporalityStartTimes(t *testing.T) {
 
 			// Add debug logging before the check
 			t.Logf("Timestamp debug logging:\n"+
-				"First start time:  %v (%d UnixNano)\n"+
-				"Second start time: %v (%d UnixNano)\n"+
+				"First start time:  %s\n"+
+				"Second start time: %s\n"+
 				"Difference: %v\n"+
 				"Temporality: %v",
-				firstStartTime.Format(time.RFC3339Nano),
-				firstStartTime.UnixNano(),
-				secondStartTime.Format(time.RFC3339Nano),
-				secondStartTime.UnixNano(),
+				firstStartTime.String(),
+				secondStartTime.String(),
 				secondStartTime.Sub(firstStartTime),
 				tt.temporality)
 

--- a/cmd/telemetrygen/pkg/metrics/worker_test.go
+++ b/cmd/telemetrygen/pkg/metrics/worker_test.go
@@ -486,18 +486,20 @@ func TestTemporalityStartTimes(t *testing.T) {
 			name:        "Cumulative temporality keeps same start timestamp",
 			temporality: AggregationTemporality(metricdata.CumulativeTemporality),
 			checkTimes: func(t *testing.T, firstTime, secondTime time.Time) {
-				assert.Equal(t, firstTime, secondTime,
-					"cumulative metrics should have same start time")
+				if !assert.Equal(t, firstTime, secondTime,
+					"cumulative metrics should have same start time") {
+					logTimestampDiff(t, firstTime, secondTime)
+				}
 			},
 		},
 		{
 			name:        "Delta temporality has different start timestamps",
 			temporality: AggregationTemporality(metricdata.DeltaTemporality),
 			checkTimes: func(t *testing.T, firstTime, secondTime time.Time) {
-				diff := secondTime.Sub(firstTime)
-				t.Logf("Time difference: %v", diff)
-				assert.True(t, secondTime.After(firstTime),
-					"delta metrics should have increasing start times")
+				if !assert.True(t, secondTime.After(firstTime),
+					"delta metrics should have increasing start times") {
+					logTimestampDiff(t, firstTime, secondTime)
+				}
 			},
 		},
 	}
@@ -539,18 +541,17 @@ func TestTemporalityStartTimes(t *testing.T) {
 			firstStartTime := firstMetric.Data.(metricdata.Sum[int64]).DataPoints[0].StartTime
 			secondStartTime := secondMetric.Data.(metricdata.Sum[int64]).DataPoints[0].StartTime
 
-			// Add debug logging before the check
-			t.Logf("Timestamp debug logging:\n"+
-				"First start time:  %s\n"+
-				"Second start time: %s\n"+
-				"Difference: %v\n"+
-				"Temporality: %v",
-				firstStartTime.String(),
-				secondStartTime.String(),
-				secondStartTime.Sub(firstStartTime),
-				tt.temporality)
-
 			tt.checkTimes(t, firstStartTime, secondStartTime)
 		})
 	}
+}
+
+func logTimestampDiff(t *testing.T, firstTime, secondTime time.Time) {
+	t.Logf("Timestamp debug logging:\n"+
+		"First start time:  %s\n"+
+		"Second start time: %s\n"+
+		"Difference: %v",
+		firstTime.String(),
+		secondTime.String(),
+		secondTime.Sub(firstTime))
 }

--- a/cmd/telemetrygen/pkg/metrics/worker_test.go
+++ b/cmd/telemetrygen/pkg/metrics/worker_test.go
@@ -532,6 +532,19 @@ func TestTemporalityStartTimes(t *testing.T) {
 			firstStartTime := firstMetric.Data.(metricdata.Sum[int64]).DataPoints[0].StartTime
 			secondStartTime := secondMetric.Data.(metricdata.Sum[int64]).DataPoints[0].StartTime
 
+			// Add debug logging before the check
+			t.Logf("Timestamp debug logging:\n"+
+				"First start time:  %v (%d UnixNano)\n"+
+				"Second start time: %v (%d UnixNano)\n"+
+				"Difference: %v\n"+
+				"Temporality: %v",
+				firstStartTime.Format(time.RFC3339Nano),
+				firstStartTime.UnixNano(),
+				secondStartTime.Format(time.RFC3339Nano),
+				secondStartTime.UnixNano(),
+				secondStartTime.Sub(firstStartTime),
+				tt.temporality)
+
 			tt.checkTimes(t, firstStartTime, secondStartTime)
 		})
 	}

--- a/cmd/telemetrygen/pkg/metrics/worker_test.go
+++ b/cmd/telemetrygen/pkg/metrics/worker_test.go
@@ -5,7 +5,6 @@ package metrics
 
 import (
 	"context"
-	"runtime"
 	"sync"
 	"sync/atomic"
 	"testing"
@@ -478,9 +477,6 @@ func configWithMultipleAttributes(metric MetricType, qty int) *Config {
 }
 
 func TestTemporalityStartTimes(t *testing.T) {
-	if runtime.GOOS == "windows" {
-		t.Skip("http://github.com/open-telemetry/opentelemetry-collector-contrib/issues/39219")
-	}
 	tests := []struct {
 		name        string
 		temporality AggregationTemporality

